projects/observability/src/shared/components/cartesian/d3/legend/cartesian-legend.ts

```diff
@@ -202,16 +260,57 @@ export class CartesianLegend {
     this.activeSeriesSubject.next();
   }
 
-  private updateActiveSeries(seriesEntry: Series<{}>): void {
-    if (this.isDefault) {
-      this.activeSeries = [];
-      this.activeSeries.push(seriesEntry);
-      this.isDefault = false;
+  private getGroupedSeries(): Series<{}>[][] {
+    if (!this.isGrouped) {
+      return [];
+    }
+
+    const nonHiddenSeries = this.series.filter(series => !series.hide);
+    const groupedSeries: Series<{}>[][] = [];
+    const currentGroup: Series<{}>[] = [];
+
+    nonHiddenSeries.forEach(seriesEntry => {
+      if (
+        currentGroup.length === 0 ||
+        currentGroup[currentGroup.length - 1].assignedGroup!.id === seriesEntry.assignedGroup!.id
+      ) {
+        currentGroup.push(seriesEntry);
```

Comment:
Functional is your friend here - you should be able to use utils like `groupBy`  to quickly and more importantly readably, reshape an array into a map.